components/script/dom/compressionstream.rs

```diff
@@ -0,0 +1,277 @@
+    compressor
+        .write_all(&chunk)
+        .map_err(|_| Error::Type("Fail to write all to compressor".to_string()))?;
+    compressor
+        .flush()
+        .map_err(|_| Error::Type("Fail to flush compressor".to_string()))?;
+    let buffer = &compressor.get_ref()[offset..];
+
+    // Step 3. If buffer is empty, return.
+    if buffer.is_empty() {
+        return Ok(());
+    }
+
+    // Step 4. Let arrays be the result of splitting buffer into one or more non-empty pieces and
+    // converting them into Uint8Arrays.
+    // Step 5. For each Uint8Array array of arrays, enqueue array in cs’s transform.
+    // NOTE: We process the result in a single Uint8Array.
+    rooted!(in(*cx) let mut js_object = ptr::null_mut::<JSObject>());
+    let buffer_source: Uint8Array =
+        create_buffer_source(cx, buffer, js_object.handle_mut(), can_gc)
+            .map_err(|_| Error::Type("Cannot convert byte sequence to Uint8Array".to_owned()))?;
+    rooted!(in(*cx) let mut rval = UndefinedValue());
+    buffer_source.safe_to_jsval(cx, rval.handle_mut());
+    controller.enqueue(cx, global, rval.handle(), can_gc)?;
+
+    // NOTE: We don't need to keep result that has been copied to Uint8Array. Clear the inner
+    // buffer of compressor to save memory.
+    compressor.get_mut().clear();
+
+    Ok(())
+}
+
+/// <https://compression.spec.whatwg.org/#compress-flush-and-enqueue>
+pub(crate) fn compress_flush_and_enqueue(
+    cx: SafeJSContext,
+    global: &GlobalScope,
+    cs: &CompressionStream,
+    controller: &TransformStreamDefaultController,
+    can_gc: CanGc,
+) -> Fallible<()> {
+    // Step 1. Let buffer be the result of compressing an empty input with cs’s format and context,
+    // with the finish flag.
+    // NOTE: In our implementation, the enum type of context already indicates the format.
+    let mut compressor = cs.context.borrow_mut();
+    let offset = compressor.get_ref().len();
+    compressor
+        .try_finish()
+        .map_err(|_| Error::Type("Fail to finish compressor".to_string()))?;
+    let buffer = &compressor.get_ref()[offset..];
+
+    // Step 2. If buffer is empty, return.
+    if buffer.is_empty() {
+        return Ok(());
+    }
+
+    // Step 3. Let arrays be the result of splitting buffer into one or more non-empty pieces and
+    // converting them into Uint8Arrays.
+    // Step 4. For each Uint8Array array of arrays, enqueue array in cs’s transform.
+    // NOTE: We process the result in a single Uint8Array.
+    rooted!(in(*cx) let mut js_object = ptr::null_mut::<JSObject>());
+    let buffer_source: Uint8Array =
+        create_buffer_source(cx, buffer, js_object.handle_mut(), can_gc)
+            .map_err(|_| Error::Type("Cannot convert byte sequence to Uint8Array".to_owned()))?;
+    rooted!(in(*cx) let mut rval = UndefinedValue());
+    buffer_source.safe_to_jsval(cx, rval.handle_mut());
+    controller.enqueue(cx, global, rval.handle(), can_gc)?;
+
+    Ok(())
```

Comment:
You already clear the compressor’s buffer after each transformed chunk to save memory you probably shoudl Do the same in the flush path to avoid retaining the final output slice?!